src/util/Timer.cpp

```diff
@@ -139,63 +146,47 @@ VirtualClock::enqueue(VirtualClockEvent const& ve)
     mEvents.emplace(ve);
 }
 
-bool
-VirtualClock::cancelAllEventsFrom(VirtualTimer& v)
+void
+VirtualClock::flushCancelledEvents()
 {
     if (mDestructing)
     {
-        return false;
+        return;
+    }
+    if (mFlushesIgnored <= mEvents.size())
+    {
+        mFlushesIgnored++;
+        return;
     }
     // LOG(DEBUG) << "VirtualClock::cancelAllEventsFrom";
-    // Brute force approach; could be done better with a linked list
-    // per timer, as is done in asio. For the time being this should
-    // be tolerable.
-    vector<VirtualClockEvent> toCancel;
-    vector<VirtualClockEvent> toKeep;
-    toCancel.reserve(mEvents.size());
-    toKeep.reserve(mEvents.size());
-    bool changed = false;
+
+    auto toKeep = vector<shared_ptr<VirtualClockEvent>>();
+
     while (!mEvents.empty())
     {
         auto const& e = mEvents.top();
-        if (e.live())
+        if (!e->getTriggered())
         {
-            if (e.mTimer == &v)
-            {
-                changed = true;
-                toCancel.emplace_back(e);
-            }
-            else
-            {
-                toKeep.emplace_back(e);
-            }
-        }
-        else
-        {
-            changed = true;
+            toKeep.emplace_back(e);
         }
         mEvents.pop();
     }
-    for (auto& e : toCancel)
-    {
-        e.mCallback(asio::error::operation_aborted);
-    }
-    mEvents = priority_queue<VirtualClockEvent>(toKeep.begin(), toKeep.end());
-    return changed;
+    mFlushesIgnored = 0;
+    mEvents = PrQueue(toKeep.begin(), toKeep.end());
+    return;
 }
 
 bool
 VirtualClock::cancelAllEvents()
 {
-    bool empty = mEvents.empty();
-    auto events = mEvents;
-    mEvents = priority_queue<VirtualClockEvent>();
-    while (!events.empty())
+    bool wasEmpty = mEvents.empty();
+    while (!mEvents.empty())
     {
-        events.top().mCallback(asio::error::operation_aborted);
-        events.pop();
+        mEvents.top()->cancel();
```

Comment:
Because you're not pulling the events aside anymore, the top element of mEvents can change between `cancel()` and `pop()`, which will mean dropping events.